README.md

```diff
@@ -3,58 +3,153 @@ xqueue_watcher
+from XQueue. There is a base grader defined in xqueue_watcher: Grader and JailedGrader (for Python, using CodeJail). If you don't use JailedGrader, you'd have to implement your own Grader by subclassing `xqueue_watcher.grader.Grader
+
+The payload from XQueue will be a JSON that usually looks like this, notice that "grader" is a required field in the "grader_payload" and must be configured accordingly in the Studio for the exercise.
+```json
+{
+    "student_info": {
+        "random_seed": 1,
+        "submission_time": "20210109222647",
+        "anonymous_student_id": "6d07814a4ece5cdda54af1558a6dfec0"
+    },
+    "grader_payload": "\n        {\"grader\": \"relative/path/to/grader.py\"}\n      ",
+    "student_response": "print \"hello\"\r\n      "
+}
+```
+
+## Custom Handler
 To implement a pull grader:
 
-Subclass xqueue_watcher.grader.Grader and override the `grade` method. Then add your grader to the config like `"handler": "my_module.MyGrader"`. The arguments for the `grade` method are:
-	* `grader_path`: absolute path to the grader defined for the current problem
-	* `grader_config`: other configuration particular to the problem
-	* `student_response`: student-supplied code
+Subclass `xqueue_watcher.grader.Grader` and override the `grade` method. Then add your grader to the config like `"handler": "my_module.MyGrader"`. The arguments for the `grade` method are:
+   * `grader_path`: absolute path to the grader defined for the current problem.
+   * `grader_config`: other configuration particular to the problem
+   * `student_response`: student-supplied code
 
+Note that `grader_path` is constructed by appending the relative path to the grader from `grader_payload` to the `grader_root` in the configuration JSON. If the handler cannot find a `grader.py` file, it would fail to grade the submission.
 
-Sandboxing
-==========
-To sandbox python, use [CodeJail](https://github.com/edx/codejail). In your handler configuration, add:
+## Grading Python submissions with JailedGrader
 
-	"CODEJAIL": {
-		"name": "python",
-		"python_bin": "/path/to/sandbox/python",
-		"user": "sandbox_username"
-	}
+`xqueue_watcher` provides a few utilities for grading python submissions, including JailedGrader for running python code in a safe environment and grading support utilities.
 
+### JailedGrader
+To sandbox python, use [CodeJail](https://github.com/edx/codejail). In your handler configuration, add:
+```json
+    "HANDLER": "xqueue_watcher.jailedgrader.JailedGrader",
+    "CODEJAIL": {
+        "name": "python",
+        "python_bin": "/path/to/sandbox/python",
+        "user": "sandbox_username"
+    }
+```
 Then, `codejail_python` will automatically be added to the kwargs for your handler. You can then import codejail.jail_code and run `jail_code("python", code...)`. You can define multiple sandboxes and use them as in `jail_code("special-python", ...)`
```

Comment:
what does `special-python` mean? This last sentence confuses me.

Reply:
Maybe `special-python` -> `python-version1`/`another-python-version`?